blockly-ws-search/src/WorkspaceSearch.js

```diff
@@ -0,0 +1,658 @@
+   */
+  init() {
+    injectSearchCss();
+    this.createDom_();
+    this.setVisible_(false);
+  }
+
+  /**
+   * Creates and injects the search bar's DOM.
+   * @protected
+   */
+  createDom_() {
+    /*
+     * Creates the search bar. The generated search bar looks like:
+     * <div class="ws-search'>
+     *   <div class="ws-search-container'>
+     *     <div class="ws-search-content'>
+     *       <div class="ws-search-input'>
+     *         [... text input goes here ...]
+     *       </div>
+     *       [... actions div goes here ...]
+     *     </div>
+     *     [... close button goes here ...]
+     *   </div>
+     * </div>
+     */
+    const injectionDiv = this.workspace_.getInjectionDiv();
+    injectionDiv.addEventListener('keydown',
+        evt => this.onWorkspaceKeyDown_(/** @type {KeyboardEvent} */ evt));
+
+    this.htmlDiv_ = document.createElement('div');
+    Blockly.utils.dom.addClass(this.htmlDiv_, 'blockly-ws-search');
+    this.positionSearchBar_();
+
+    const searchContainer = document.createElement('div');
+    Blockly.utils.dom.addClass(searchContainer, 'blockly-ws-search-container');
+
+    const searchContent = document.createElement('div');
+    Blockly.utils.dom.addClass(searchContent, 'blockly-ws-search-content');
+    searchContainer.appendChild(searchContent);
+
+    const inputWrapper = document.createElement('div');
+    Blockly.utils.dom.addClass(inputWrapper, 'blockly-ws-search-input');
+    this.inputElement_ = this.createTextInput_();
+    this.inputElement_.addEventListener('keydown',
+        evt => this.onKeyDown_(/** @type {KeyboardEvent} */ evt));
+    this.inputElement_.addEventListener('input', () => this.onInput_());
+    this.inputElement_.addEventListener('click',
+        () => this.searchAndHighlight(this.searchText_, this.preserveSelected));
+    inputWrapper.appendChild(this.inputElement_);
+    searchContent.appendChild(inputWrapper);
+
+    this.actionDiv_ = document.createElement('div');
+    Blockly.utils.dom.addClass(this.actionDiv_, 'blockly-ws-search-actions');
+    searchContent.appendChild(this.actionDiv_);
+
+    const nextBtn = this.createNextBtn_();
+    if (nextBtn) {
+      this.addActionBtn(nextBtn, () => this.next());
+    }
+
+    const previousBtn = this.createPreviousBtn_();
+    if (previousBtn) {
+      this.addActionBtn(previousBtn, () => this.previous());
+    }
+
+    const closeBtn = this.createCloseBtn_();
+    if (closeBtn) {
+      this.addBtnListener_(closeBtn, () => this.close());
+      searchContainer.appendChild(closeBtn);
+    }
+
+    this.htmlDiv_.appendChild(searchContainer);
+
+    injectionDiv.insertBefore(this.htmlDiv_, this.workspace_.getParentSvg());
+  }
+
+  /**
+   * Disposes of workspace search.
+   * Unlink from all DOM elements to prevent memory leaks.
+   */
+  dispose() {
+    if (this.htmlDiv_) {
+      this.htmlDiv_.remove();
+      this.htmlDiv_ = null;
+    }
+    this.actionDiv_ = null;
+    this.inputElement_ = null;
```

Comment:
I have a couple thoughts about this: 
- Should this also call workspace_.dispose as workspace search is usually wrapped around a workspace.
- You aren't disposing of event handlers that you've registered.

Reply:
done